<launch>

    <!--==========================-->
    <!-- Arguments and Parameters -->
    <!--==========================-->
    <arg name="manual_deadman" default="4" />
    <arg name="autonomous_deadman" default="5" />
    <arg name="timeout" default="1.0" />
    <arg name="port_steering" default="/dev/ttyACM0"/>
    <arg name="port_accelerator_encoder" default="/dev/ttyACM1" />
    <arg name="port_clamp" default="/dev/ttyACM2"/>
    <arg name="output_max" default="255"/>
    <arg name="output_min" default="0"/>
    <arg name="max_angle" default="1.309"/> <!-- 75deg -->
    <arg name="min_angle" default="-1.309"/> <!-- -75deg -->
    <arg name="max_vel_scale" default="0.9"/>
    <arg name="ramp_time_vel" default="1.5"/>

    <!--========================-->
    <!-- Steering Control Nodes -->
    <!--========================-->
    <include file="$(find sensors)/launch/localization_steer.launch">
        <arg name="port_steering" value="$(arg port_steering)"/>
    </include>
    <node pkg="rosserial_python" type="serial_node.py" name="steering_node_A4988_arduino">
        <param name="port" value="/dev/ttyACM2"/>
    </node>
    <node pkg="motion_testing" type="steering_node_A4988.py" name="steering_node" respawn="true">
        <param name="manual_deadman" value="$(arg manual_deadman)" />
        <param name="autonomous_deadman" value="$(arg autonomous_deadman)" />
        <param name="timeout" value="$(arg timeout)" />
        <param name="max_vel_scale" value="$(arg max_vel_scale)"/>
        <param name="ramp_time_vel" value="$(arg ramp_time_vel)"/>
    </node>
    
    <!--========================-->
    <!-- Velocity Control Nodes -->
    <!--========================-->
    <!--
    NOTE: the canbus velocity reading and IMU data should be collected
          through the Raspberry Pi by running the following commands:
          1) $ sudo slcan_add.sh (this starts the canbus reader)
          2) $ roslaunch sensors localization_velocity_imu.launch (reads imu and
               canbus data)
    -->
    <!-- The arduino running this code must be a MEGA and should be running "motion_testing/src/accelerator_encoder_gear_analog.ino", see file for wiring -->
    <node pkg="rosserial_python" type="serial_node.py" name="accelerator_encoder_gear_arduino">
        <param name="port" value="$(arg port_accelerator_encoder)"/>
    </node>
    <node pkg="motion_testing" type="velocity_gear_two_pid_delta.py" name="velocity_gear_two_pid_delta">
        <param name="manual_deadman" value="$(arg manual_deadman)"/>
        <param name="autonomous_deadman" value="$(arg autonomous_deadman)"/>
        <param name="timeout" value="$(arg timeout)"/>
        <param name="Kp1" value="1.5"/>
        <param name="Ki1" value="0"/>
        <param name="Kd1" value="0"/>
        <param name="Kp2" value="1.0"/>
        <param name="Ki2" value="0"/>
        <param name="Kd2" value="0"/>
        <param name="delta_max" value="100"/>
        <param name="delta_min" value="-100"/>
        <param name="error_tolerance" value="0.05"/>
        <param name="error_exponent1" value="1.0" />
        <param name="error_exponent2" value="1.0" />
        <param name="output_max" value="$(arg output_max)"/>
        <param name="output_min" value="$(arg output_min)"/>
    </node>

</launch>
